src/fullnode.rs

```diff
@@ -213,65 +225,146 @@ impl Fullnode {
-                    entry_height,
-                    blob_recycler.clone(),
-                    blob_receiver,
-                );
-                thread_hdls.extend(broadcast_stage.thread_hdls());
+                node_role = NodeRole::Leader;
+            }
+        }
+
+        let keypair = Arc::new(keypair);
+        let tvu = Tvu::new(
+            keypair.clone(),
+            &bank,
+            entry_height,
+            crdt.clone(),
+            window.clone(),
+            node.sockets.replicate,
+            node.sockets.repair,
+            node.sockets.retransmit,
+            ledger_path,
+            exit.clone(),
+            node_role == NodeRole::Leader,
+        );
+
+        let tick_duration = None;
+        // TODO: To light up PoH, uncomment the following line:
+        //let tick_duration = Some(Duration::from_millis(1000));
+
+        let tpu_option;
+        if let Some(ledger_path) = ledger_path {
+            let (tpu, blob_receiver) = Tpu::new(
+                keypair.clone(),
+                &bank,
+                &crdt,
+                tick_duration,
+                node.sockets.transaction,
+                &blob_recycler,
+                exit.clone(),
+                ledger_path,
+                sigverify_disabled,
+                node_role == NodeRole::Validator,
+            );
+
+            let broadcast_stage = BroadcastStage::new(
+                node.sockets.broadcast,
+                crdt.clone(),
+                window,
+                entry_height,
+                blob_recycler.clone(),
+                blob_receiver,
+            );
+
+            thread_hdls.extend(broadcast_stage.thread_hdls());
+            tpu_option = Some(tpu);
+        } else {
+            tpu_option = None;
+        }
+
+        Fullnode {
+            exit,
+            thread_hdls,
+            node_role,
+            id: keypair.pubkey(),
+            crdt,
+            tpu: tpu_option,
+            tvu,
+            bank,
+        }
+    }
+
+    pub fn handle_new_leader(&mut self, new_leader_id: Pubkey) {
+        self.crdt
+            .write()
+            .unwrap()
+            .update_leader(Some(new_leader_id));
+        if self.node_role == NodeRole::Leader && new_leader_id != self.id {
+            self.transition_role(NodeRole::Validator)
+        }
+
+        if self.node_role == NodeRole::Validator && new_leader_id == self.id {
+            self.transition_role(NodeRole::Leader)
+        }
+    }
+
+    fn transition_role(&mut self, new_role: NodeRole) {
+        match (self.node_role, new_role) {
+            (NodeRole::Leader, NodeRole::Validator) => {
+                self.bank.is_leader.store(false, Ordering::Relaxed);
+                self.start_validator();
+                self.node_role = NodeRole::Validator;
+            }
+            (NodeRole::Validator, NodeRole::Leader) => {
+                // If this is a validator only node (didn't supply a ledger path), 
```

Comment:
the absence of a ledger_path in Fullnode::new() was to accommodate tests that existed before validators were writing to persistent ledgers.

there are 2 paths in the current code to writing the ledger in the current code base...  we need a test that a ledger written by a node that does transitions is still a good ledger...  something like that could be covered by a ledger test, or moving LedgerWriter to a share-able value shared among all stages instead of having 2 LedgerWriters